packages/eslint-plugin/src/rules/shorthand-property-sorting/index.ts

```diff
@@ -0,0 +1,140 @@
+            a.type === 'Property' &&
+            a.key.type === 'Identifier' &&
+            b.type === 'Property' &&
+            b.key.type === 'Identifier'
+          ) {
+            const propA = kebabCase(a.key.name) as ShorthandProperties;
+            const propB = kebabCase(b.key.name) as ShorthandProperties;
+
+            return shorthandBuckets[propA] - shorthandBuckets[propB];
+          }
+          return 0;
+        });
+
+        const sourceCode = context.getSourceCode();
+        const sortedCode = sortedProperties
+          .map((property) => sourceCode.getText(property))
+          .join(', ');
+
+        // Replace the old object expression with the new sorted one
+        const newObjectExpression = `{ ${sortedCode} }`;
+
+        return fixer.replaceText(node, newObjectExpression);
+      }
+      return null;
+    },
+  });
+};
+
+const callExpressionCorrectImport = (node: Rule.Node, references: Scope.Reference[]): boolean => {
+  return (
+    node.type === 'CallExpression' &&
+    (isCss(node.callee as Rule.Node, references) ||
+      isCssMap(node.callee as Rule.Node, references) ||
+      isStyled(node.callee as Rule.Node, references))
+  );
+};
+
+export const shorthandFirst: Rule.RuleModule = {
+  meta: {
+    docs: {
+      recommended: true,
+      description:
+        "At build time, Compiled automatically sorts shorthand properties (like `font` and `border`) so that they come before any longhand properties (like `fontSize` and `borderTopColor`) defined on the component. This means that longhand properties will always override shorthand properties. This rule enforces that the order in which the properties appear in a component's source code matches the actual ordering the properties will have at build time and runtime.",
+      url: 'https://github.com/atlassian-labs/compiled/tree/master/packages/eslint-plugin/src/rules/shorthand-property-sorting',
+    },
+    messages: {
+      'shorthand-first':
+        'When using both shorthand and longhand properties, the shorthand property should be first.',
+    },
+    type: 'suggestion',
+    fixable: 'code',
+    hasSuggestions: true,
+  },
+  create(context) {
+    const selectorString: string =
+      'CallExpression[callee.type="Identifier"][callee.name="css"] > ObjectExpression,' +
+      'CallExpression[callee.type="Identifier"][callee.name="cssMap"] > ObjectExpression,' +
+      'CallExpression[callee.type="Identifier"][callee.name="xcss"] > ObjectExpression,' +
+      'CallExpression[callee.type="MemberExpression"][callee.object.name="styled"] > ObjectExpression';
+
+    return {
+      [selectorString]: (node: Rule.Node) => {
+        if (node.type === 'ObjectExpression' && node.properties.length > 0) {
+          let lowestDepth = 0;
+          let fixRequired = false;
+
+          const references = context.getScope().references;
+
+          // loop through the css properties of a ObjectExpression object
+          node.properties.some((property) => {
```

Comment:
I haven't worked with Compiled in a while, but rather than an ESLint rule, I wonder whether we could have leverage the existing logic and throw an error at compilation time instead. Is doing this with ESLint more efficient/performant and less disruptive, since linting can be optional and it won't affect build time?

cc @dddlr

Reply:
we could yeah, and it would definitely be more robust (ESLint doesn't let us detect violations across components, but a compile time approach might)

but a compile time approach would be harder (and we'd need to make exceptions for the few times that DST deliberately puts longhand properties before shorthand properties), and considering that it's hard to justify impact, I'm hesitant to give ourselves more work there

Reply:
curious to hear what you think - i could be wrong or missed something

Reply:
I am hesitant too because I haven't worked with Compiled in a while - however, I got the feeling that we could just throwing an error when generating the classes. Also, I don't know what implications could cause and I am hesitant on adding changes that could impact build times